joblib/numpy_pickle.py

```diff
@@ -63,135 +47,422 @@ def _read_magic(file_handle):
     return magic
 
 
-def read_zfile(file_handle):
-    """Read the z-file and return the content as a string
+def _check_filetype(filename, magic):
+    """ Utility function opening and returning a fileobject from a filename
+    given it's magic number (compressed using gzip or not compressed
+
+    Parameters
+    ----------
+    filename: str
+        The full path string of the file
+    magic: bytes
+        The bytes representation of the magic number of the file
+
+    Returns
+    -------
+        a file like object
 
-    Z-files are raw data compressed with zlib used internally by joblib
-    for persistence. Backward compatibility is not guaranteed. Do not
-    use for external purposes.
     """
-    file_handle.seek(0)
-    assert _read_magic(file_handle) == _ZFILE_PREFIX, \
-        "File does not have the right magic"
-    header_length = len(_ZFILE_PREFIX) + _MAX_LEN
-    length = file_handle.read(header_length)
-    length = length[len(_ZFILE_PREFIX):]
-    length = int(length, 16)
-
-    # With python2 and joblib version <= 0.8.4 compressed pickle header is one
-    # character wider so we need to ignore an additional space if present.
-    # Note: the first byte of the zlib data is guaranteed not to be a
-    # space according to
-    # https://tools.ietf.org/html/rfc6713#section-2.1
-    next_byte = file_handle.read(1)
-    if next_byte != b' ':
-        # The zlib compressed data has started and we need to go back
-        # one byte
-        file_handle.seek(header_length)
-
-    # We use the known length of the data to tell Zlib the size of the
-    # buffer to allocate.
-    data = zlib.decompress(file_handle.read(), 15, length)
-    assert len(data) == length, (
-        "Incorrect data length while decompressing %s."
-        "The file could be corrupted." % file_handle)
-    return data
-
-
-def write_zfile(file_handle, data, compress=1):
-    """Write the data in the given file as a Z-file.
-
-    Z-files are raw data compressed with zlib used internally by joblib
-    for persistence. Backward compatibility is not guarantied. Do not
-    use for external purposes.
+    if magic == _GZIP_PREFIX:
+        return gzip.GzipFile(filename, 'rb')
+
+    return open(filename, 'rb')
+
+
+# difference between version 1.0 and 2.0 is a 4 byte (I) header length
+# instead of 2 bytes (H) allowing storage of large structured arrays
+
+def _check_version(version):
+    """
+    Utility that check if the version is supported. An exception is raised if
+    version is not supported.
+    Parameters
+    ----------
+    version: tuple
+        2 element tuple containing the major and minor versions.
+
+    Raises
+    ------
+    ValueError
+        If the version is invalid.
+
+    """
+    if version not in [(1, 0), (2, 0), None]:
+        msg = "we only support format version (1,0) and (2, 0), not %s"
+        raise ValueError(msg % (version,))
+
+
+def _read_bytes(fp, size, error_template="ran out of data"):
```

Comment:
You needed to copy and paste some code from numpy IIRC. Can you remind me why?

It' be great if it was labeled clearly as such in each function docstring and you mentioned which version of numpy you took it from.